pkg/socketlb/cgroup.go

```diff
@@ -0,0 +1,240 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package socketlb
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"golang.org/x/sys/unix"
+)
+
+var attachTypes = map[string]ebpf.AttachType{
+	Connect4:     ebpf.AttachCGroupInet4Connect,
+	SendMsg4:     ebpf.AttachCGroupUDP4Sendmsg,
+	RecvMsg4:     ebpf.AttachCGroupUDP4Recvmsg,
+	GetPeerName4: ebpf.AttachCgroupInet4GetPeername,
+	PostBind4:    ebpf.AttachCGroupInet4PostBind,
+	PreBind4:     ebpf.AttachCGroupInet4Bind,
+	Connect6:     ebpf.AttachCGroupInet6Connect,
+	SendMsg6:     ebpf.AttachCGroupUDP6Sendmsg,
+	RecvMsg6:     ebpf.AttachCGroupUDP6Recvmsg,
+	GetPeerName6: ebpf.AttachCgroupInet6GetPeername,
+	PostBind6:    ebpf.AttachCGroupInet6PostBind,
+	PreBind6:     ebpf.AttachCGroupInet6Bind,
+}
+
+// attachCgroup attaches a program from spec with the given name to cgroupRoot.
+// If the kernel supports it, the resulting bpf_link is pinned to pinPath.
+//
+// Upgrades from prior Cilium versions are handled automatically and may detach
+// existing programs if attachment does not succeed at first. New programs are
+// attached first before detaching prior ones if possible.
+func attachCgroup(spec *ebpf.Collection, name, cgroupRoot, pinPath string) error {
+	prog := spec.Programs[name]
+	if prog == nil {
+		return fmt.Errorf("program %s not found in ELF", name)
+	}
+
+	pin := filepath.Join(pinPath, name)
+	l, err := link.LoadPinnedLink(pin, &ebpf.LoadPinOptions{})
+	if err == nil {
+		if err := l.Update(prog); err != nil {
+			return fmt.Errorf("update link %s: %w", pinPath, err)
+		}
+
+		log.Debugf("updated link %s for program %s", pinPath, name)
+		return nil
+	}
+	log.Debugf("no pinned link at '%s'", pinPath)
+
+	l, err = link.AttachCgroup(link.CgroupOptions{
+		Path:    cgroupRoot,
+		Attach:  attachTypes[name],
+		Program: prog,
+	})
+
+	switch {
+	// Hit a recoverable error that could be resolved by detaching all programs
+	// from the cgroup hook.
+	//
+	// Kernels before 5.5 don't support PROG_ATTACH atomic replace.
+	// link.AttachCgroup returns ErrNotSupported when a PROG_ATTACH prog is active
+	// and it cannot be replaced without removing it first.
+	//
+	// If the kernel supports bpf_link, and an older version of Cilium attached a
+	// cgroup program without flags, link.AttachCgroup will return EPERM because
+	// bpf_link implicitly uses the multi flag.
+	case errors.Is(err, unix.EPERM) || errors.Is(err, link.ErrNotSupported):
+		log.Debugf("performing exclusive attach for program %s", name)
+
+		l, err = exclusiveCgroupAttach(prog, name, cgroupRoot)
+		if err != nil {
+			return fmt.Errorf("exclusively attach program %s: %w", name, err)
+		}
```

Comment:
Replying because it will be relevant even after this set gets re-spun: when this function (`attachCgroup()`) enters, all entrypoints in `bpf_sock.o` have already been verified and loaded. `link.AttachCgroup()` doesn't return verifier errors.

Regarding migrations: both the old (init.sh) and new code share the same flaw that if cgroup attachment fails partway through (some hooks get replaced, some don't), we'll be in a split-brain situation regardless. This is impossible to fully defend against; we could retain the handles of all 'old' programs we replace and reinstate them if any errors are encountered, but if the agent receives a SIGKILL at any point, that falls over as well.

Reply:
IMO the problem isn't so much that it complicates the code, but rather that there's no sane degradation path for the way the agent is currently designed. Also, consider for a moment that prog replacements (on cgroups, specifically) failing partway through should be exceedingly rare, with the only cases I can think of being caused by bugs in Cilium. (attaching a program of the wrong attach type, etc.)

You mention `Ah, load/attach/etc. for program X failed, but Cilium left the system in a functional state and now I can investigate ...`, but under no circumstance should agent startup continue if cgroup attachment failed, since subsequent modules can attach the new datapath while the old cgroup progs are still attached.

A case could be made for rolling back programs, since that puts the node back into its previously-known configuration. But what if other subsystems already successfully attached the 'new' datapath? Those would have to be rolled back as well. We'd have to consider all programs attached during agent startup as part of a single transaction that potentially needs to be rolled back. This would require all subsystems that attach bpf progs to first scan all programs they would potentially replace, take out references to them so they don't get deallocated, then perform the replace operation. Then, when all subsystems have initialized successfully, forget all fds of the old datapath.

I think we're quite far out from this being a reality, let alone tackling that in this PR. It'll probably require some kind of central bpf manager component that can track datapath revisions and implement rollbacks that way. And then of course, the question remains: what if rolling back also fails? We end up in a rollback scenario due to a failed load/attach to begin with; who's to say the rollback will even succeed? Now the Cilium agent enters a restart loop where it'll fail to attach some new program, and keep rolling back to the partially-updated dataplane it found during startup.

Hence my conclusion of this not being worth it. If a node ends up in a state like this, there is nothing left to do but nuke at least part of it. Manual intervention is necessary anyway.